src/wat-writer.cc

```diff
@@ -807,9 +871,46 @@ void WatWriter::PushExpr(const Expr* expr,
     expr_tree_stack_.emplace_back(expr);
     FlushExprTreeStack();
   }
+  if (s_trace) {
+    fprintf(stderr, "-> PushExpr(%s, %" PRIindex ", %" PRIindex")\n",
+            GetExprTypeName(*expr), operand_count, result_count);
+  }
+}
+
+void WatWriter::PushCatch(const Catch* catch_) {
+  if (s_trace)
+    fprintf(stderr, "-> PushCatch()\n");
+  ExprTree tree(catch_);
+  expr_tree_stack_.push_back(std::move(tree));
+  FlushExprTreeStack();
+  if (s_trace)
+    fprintf(stderr, "<- PushCatch()\n");
+}
+
+void WatWriter::FlushCatchTree(const ExprTree& expr_tree) {
```

Comment:
That's fair... :-) Probably could use a comment or two explaining what's going on too. I'll see about adding something like that... the basic idea is relatively straightforward (I think!)